great_expectations/datasource/fluent/interfaces.py

```diff
@@ -536,22 +560,40 @@ def _execution_engine_type(self) -> Type[_ExecutionEngineT]:
         return self.execution_engine_override or self.execution_engine_type
 
     def add_batch_config(self, batch_config: BatchConfig) -> BatchConfig:
+        asset_name = batch_config.data_asset.name
         if not self.data_context:
-            raise DataContextError("Datasource is not attached to a DataContext")
-        return self.data_context.datasource_store.add_batch_config(batch_config)
+            raise DataContextError("Cannot save datasource without a data context.")
+
+        loaded_datasource = self.data_context.get_datasource(self.name)
+        if loaded_datasource is not self:
+            # CachedDatasourceDict will return self; only add batch config if this is a remote copy
+            assert isinstance(loaded_datasource, Datasource)
+            loaded_asset = loaded_datasource.get_asset(asset_name)
+            loaded_asset.batch_configs.append(batch_config)
+            loaded_asset.update_batch_config_field_set()
+        updated_datasource = self.data_context.update_datasource(loaded_datasource)
```

Comment:
I need to read more about the `CachedDatasourceDict` and how that works. If there are changes to the datasource locally other than adding this batch_config do those get persisted. Or does this never happen?

Reply:
I actually wrote and then didn't commit a test around this. datasources persisted to the `CachedDatasourceDict` are the the same (as in not just `==`, but `is`). See https://github.com/great-expectations/great_expectations/blob/develop/great_expectations/datasource/datasource_dict.py#L250C14-L250C14 and https://github.com/great-expectations/great_expectations/blob/develop/great_expectations/datasource/datasource_dict.py#L268C18-L268C18